File: python/logbrew_py/src/logbrew_sdk/examples/agent_timeline.py
"""Packaged agent-readable timeline example for installed SDK users."""

from __future__ import annotations

import json
import sys

from logbrew_sdk import (
    LogBrewClient,
    RecordingTransport,
    create_network_milestone_attributes,
    create_product_action_attributes,
    create_traceparent_headers,
)


def main() -> int:
    trace_id = "4bf92f3577b34da6a3ce929d0e0e4736"
    headers = create_traceparent_headers(trace_id=trace_id, span_id="00f067aa0ba902b7")
    client = LogBrewClient.create(
        api_key="LOGBREW_API_KEY",
        sdk_name="checkout-agent-timeline",
        sdk_version="0.1.0",
    )

    client.action(
        "evt_checkout_submit",
        "2026-06-02T10:00:05Z",
        create_product_action_attributes(
            {
                "name": "checkout.submit",
                "status": "running",
                "sessionId": "sess_123",
                "traceId": trace_id,
                "routeTemplate": "/checkout/:step?email=private@example.test#payment",
                "screen": "checkout",
                "funnel": "checkout",
                "step": "submit",
                "metadata": {"service": "checkout", "payload": {"card": "redacted"}},
            }
        ),
    )
    client.action(
        "evt_payment_api",
        "2026-06-02T10:00:06Z",
        create_network_milestone_attributes(
            {
                "routeTemplate": "https://api.example.test/payments/:id?card=private#receipt",
                "method": "post",
                "statusCode": 202,
                "durationMs": 94,
                "sessionId": "sess_123",
                "traceId": trace_id,
                "metadata": {"service": "checkout", "headers": {"authorization": "redacted"}},
            }
        ),
    )

    print(client.preview_json())
    response = client.shutdown(RecordingTransport.always_accept())
    print(
        json.dumps(
            {
                "ok": True,
                "status": response.status_code,
                "attempts": response.attempts,
                "events": 2,
                "traceparent": headers["traceparent"],
            },
            sort_keys=True,
        ),
        file=sys.stderr,
    )
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: js/logbrew-js/examples/agent-timeline.mjs
const sdk = await import("@logbrew/sdk").catch(async (error) => {
  if (error && error.code === "ERR_MODULE_NOT_FOUND") {
    return import("../index.js");
  }
  throw error;
});

const {
  createNetworkMilestoneAttributes,
  createProductActionAttributes,
  createTraceparentHeaders,
  LogBrewClient,
  RecordingTransport
} = sdk;

const traceId = "4bf92f3577b34da6a3ce929d0e0e4736";
const spanId = "b7ad6b7169203331";
const sessionId = "sess_checkout_001";
const timestamp = "2026-06-02T10:00:00Z";

const client = LogBrewClient.create({
  apiKey: "LOGBREW_API_KEY",
  sdkName: "checkout-agent-timeline",
  sdkVersion: "1.0.0",
  eventFilter(event) {
    return !(event.type === "log" && event.attributes.level === "info");
  }
});

client.action("evt_checkout_started", timestamp, createProductActionAttributes({
  name: "checkout.started",
  status: "success",
  sessionId,
  traceId,
  routeTemplate: "/checkout/:step?coupon=private#payment",
  funnel: "checkout",
  step: "start",
  metadata: { service: "checkout", plan: "pro" }
}));

client.action("evt_payment_api", "2026-06-02T10:00:01Z", createNetworkMilestoneAttributes({
  routeTemplate: "https://api.example.invalid/payments/123?card=private#retry",
  method: "POST",
  statusCode: 503,
  durationMs: 241.5,
  sessionId,
  traceId,
  metadata: { service: "payments", retryable: true }
}));

client.log("evt_debug_noise", "2026-06-02T10:00:02Z", {
  message: "debug heartbeat",
  level: "info",
  logger: "checkout"
});

const headers = createTraceparentHeaders({
  traceId,
  spanId,
  traceFlags: "01"
});

const preview = client.previewJson();
if (preview.includes("card=private") || preview.includes("coupon=private") || preview.includes("#payment")) {
  throw new Error("agent timeline leaked query or hash metadata");
}
if (client.pendingEvents() !== 2) {
  throw new Error(`expected two retained timeline events, got ${client.pendingEvents()}`);
}

console.log(preview);

const response = await client.shutdown(RecordingTransport.alwaysAccept());
console.error(JSON.stringify({
  ok: true,
  events: 2,
  traceparent: headers.traceparent,
  status: response.statusCode
}));
